molgenis-js/src/main/java/org/molgenis/js/nashorn/NashornScriptEngine.java

```diff
@@ -1,107 +1,68 @@
 package org.molgenis.js.nashorn;
 
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.github.benmanes.caffeine.cache.Weigher;
 import jdk.nashorn.api.scripting.JSObject;
 import jdk.nashorn.api.scripting.NashornScriptEngineFactory;
 import jdk.nashorn.api.scripting.ScriptObjectMirror;
-import org.molgenis.util.ResourceUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.stereotype.Component;
 
 import javax.script.*;
-import java.io.IOException;
-import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
 
-import static java.util.Arrays.asList;
+import static org.molgenis.js.magma.JsMagmaScriptEvaluator.KEY_ID_VALUE;
 
 @Component
 public class NashornScriptEngine
 {
 	private static final Logger LOG = LoggerFactory.getLogger(NashornScriptEngine.class);
-
-	private static final List<String> RESOURCE_NAMES;
-
-	static
-	{
-		RESOURCE_NAMES = asList("/js/es6-shims.js", "/js/math.min.js", "/js/script-evaluator.js");
-	}
-
+	private static final int MAX_COMPILED_EXPRESSIONS_SCRIPTS_LENGTH = 500_000;
 	private ScriptEngine scriptEngine;
-	private ThreadLocal<Bindings> bindingsThreadLocal;
+	private LoadingCache<String, CompiledScript> expressions;
 
 	public NashornScriptEngine()
 	{
 		initScriptEngine();
 	}
 
-	public Object invokeFunction(String functionName, Object... args)
+	public Object eval(String script) throws ScriptException
 	{
-		Bindings bindings = bindingsThreadLocal.get();
-		Object returnValue = ((JSObject) bindings.get(functionName)).call(this, args);
-		return convertNashornValue(returnValue);
+		return convertNashornValue(scriptEngine.eval(script, new SimpleBindings()));
 	}
 
-	public ScriptObjectMirror newJSArray()
+	/**
+	 * Evaluates an expression using the given bindings.
+	 *
+	 * @param bindings   the Bindings to use as ENGINE_SCOPE
+	 * @param expression the expression to evaluate
+	 * @return result of the evaluation
+	 * @throws ScriptException if the evaluation fails
+	 */
+	public Object eval(Bindings bindings, String expression) throws ScriptException
 	{
-		Bindings bindings = bindingsThreadLocal.get();
-		return (ScriptObjectMirror) ((JSObject) bindings.get("Array")).newObject();
+		CompiledScript compiledExpression = expressions.get(expression);
+		Object returnValue = compiledExpression.eval(bindings);
+		return convertNashornValue(returnValue);
 	}
 
-	public Object eval(String script) throws ScriptException
+	public ScriptObjectMirror newJSArray()
 	{
-		Bindings bindings = bindingsThreadLocal.get();
-		return scriptEngine.eval(script, bindings);
+		return (ScriptObjectMirror) ((JSObject) scriptEngine.get("Array")).newObject();
```

Comment:
How about changing this to return List<Object> as to remove part of the magic from JsMagmaScriptEvaluator to this class?

Reply:
The ScriptObjectMirror mirrors a native javascript array. Java-side you can convert it to list to manipulate it and fill it with other objects, but when passing it back to nashorn you need to provide this ScriptObjectMirror so that it ends up as its original self and for instance Arrays.isArray evaluates to true on the javascript-side.